File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int preProd = 0, suffProd = 0;
        int res = nums[0], n = nums.size();
        for (int i = 0; i < n; i++) {
            preProd = nums[i] * (preProd == 0 ? 1 : preProd);
            suffProd = nums[n - 1 - i] * (suffProd == 0 ? 1 : suffProd);
            res = max(res, max(preProd, suffProd));
        }
        return res;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freqMap, maxF = {}, 0
        n, l, res = len(s), 0, 0
        for r in range(n):
            freqMap[s[r]] = 1 + freqMap.get(s[r], 0)
            maxF = max(maxF, freqMap[s[r]])
            while (r - l + 1 - maxF > k):
                freqMap[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res